File: tetris/src/CTetris.ts
import Matrix from "./Matrix";
import {Tetris, TetrisState} from "./Tetris";
let setOfColorBlockArrays:number[][][][] =[ // [7][4][?][?]
        [
                        [
                                        [10, 10],
                                        [10, 10]
                        ],
                        [
                                        [10, 10],
                                        [10, 10]
                        ],
                        [
                                        [10, 10],
                                        [10, 10]
                        ],
                        [
                                        [10, 10],
                                        [10, 10]
                        ]
        ],
        [
                        [
                                        [0, 20, 0],
                                        [20, 20, 20],
                                        [0, 0, 0],
                        ],
                        [
                                        [0, 20, 0],
                                        [0, 20, 20],
                                        [0, 20, 0],
                        ],
                        [
                                        [0, 0, 0],
                                        [20, 20, 20],
                                        [0, 20, 0],
                        ],
                        [
                                        [0, 20, 0],
                                        [20, 20, 0],
                                        [0, 20, 0],
                        ],
        ],
        [
                        [
                                        [30, 0, 0],
                                        [30, 30, 30],
                                        [0, 0, 0],
                        ],
                        [
                                        [0, 30, 30],
                                        [0, 30, 0],
                                        [0, 30, 0],
                        ],
                        [
                                        [0, 0, 0],
                                        [30, 30, 30],
                                        [0, 0, 30],
                        ],
                        [
                                        [0, 30, 0],
                                        [0, 30, 0],
                                        [30, 30, 0],
                        ],
        ],
        [
                        [
                                        [0, 0, 40],
                                        [40, 40, 40],
                                        [0, 0, 0],
                        ],
                        [
                                        [0, 40, 0],
                                        [0, 40, 0],
                                        [0, 40, 40],
                        ],
                        [
                                        [0, 0, 0],
                                        [40, 40, 40],
                                        [40, 0, 0],
                        ],
                        [
                                        [40, 40, 0],
                                        [0, 40, 0],
                                        [0, 40, 0],
                        ],
        ],
        [
                        [
                                        [0, 50, 0],
                                        [50, 50, 0],
                                        [50, 0, 0],
                        ],
                        [
                                        [50, 50, 0],
                                        [0, 50, 50],
                                        [0, 0, 0],
                        ],
                        [
                                        [0, 50, 0],
                                        [50, 50, 0],
                                        [50, 0, 0],
                        ],
                        [
                                        [50, 50, 0],
                                        [0, 50, 50],
                                        [0, 0, 0],
                        ],
        ],
        [
                        [
                                        [0, 60, 0],
                                        [0, 60, 60],
                                        [0, 0, 60],
                        ],
                        [
                                        [0, 0, 0],
                                        [0, 60, 60],
                                        [60, 60, 0],
                        ],
                        [
                                        [0, 60, 0],
                                        [0, 60, 60],
                                        [0, 0, 60],
                        ],
                        [
                                        [0, 0, 0],
                                        [0, 60, 60],
                                        [60, 60, 0],
                        ],
        ],
        [
                        [
                                        [0, 0, 0, 0],
                                        [70, 70, 70, 70],
                                        [0, 0, 0, 0],
                                        [0, 0, 0, 0],
                        ],
                        [
                                        [0, 70, 0, 0],
                                        [0, 70, 0, 0],
                                        [0, 70, 0, 0],
                                        [0, 70, 0, 0],
                        ],
                        [
                                        [0, 0, 0, 0],
                                        [70, 70, 70, 70],
                                        [0, 0, 0, 0],
                                        [0, 0, 0, 0],
                        ],
                        [
                                        [0, 70, 0, 0],
                                        [0, 70, 0, 0],
                                        [0, 70, 0, 0],
                                        [0, 70, 0, 0],
                        ],
        ],
];


let collBoard:Tetris = new Tetris(15,10);

class CTetris extends Tetris{



    constructor(cy:number, cx:number){
        // var bBlk = super(cy,cx);
        // var cBlk = super(cy,cx);

        super(cy,cx);
    }
    static init(setOfBlockArrays: number[][][][]){
        CTetris.nBlockTypes=setOfBlockArrays.length;
        CTetris.nBlockDegrees=setOfBlockArrays[0].length;
        CTetris.setOfBlockObjects=CTetris.createSetOfBlocks(setOfBlockArrays);
        // console.log(Tetris.setOfBlockObjects[0][0])
        // var cb:Matrix = this.setOfBlockObjects[0][0];
        // console.log(cb.print())
        CTetris.iScreenDw=CTetris.findLargestBlockSize(setOfBlockArrays);
    }

    deleteFullLines(screen:Matrix, blk:Matrix, top:number, dy:number, dx:number, dw:number){
        var line:Matrix; var zero:Matrix; var temp:Matrix;
        if(blk==null) {
            console.log("null blk")
            return screen;
        }
        var cy:number=0; var y:number=0; var nDeleted:number = 0;
        var nScanned:number = blk.get_dy();

        console.log("dfl in ct called")
        if(top+blk.get_dy()-1 >= dy)
            nScanned -= (top+blk.get_dy()-dy);
        zero=new Matrix(1,dx);
        for(y = nScanned-1; y>=0; y--){
            cy = top+y+nDeleted;
            line = screen.clip(cy,0,cy+1,screen.get_dx()).int2bool();
            line.print()


            // if(line.sum()==screen.get_dx()){

            if(line.sum()==18){
                temp = screen.clip(0,0,cy,screen.get_dx());
                screen.paste(temp,1,0);
                screen.paste(zero,0,dw);
                nDeleted++;
            }
        }
        return screen;
    }
    accept(key:string){
        // this.state = super.accept(key)
        // collBoard.accept(key)



        // console.log("collboard state: "+this.state)
        // console.log("collBoard.oScreen ; ")
        // collBoard.drawMatrix(collBoard.get_oScreen())

        var tempBlk:Matrix = new Matrix(0,0);
        // console.log("collboard state: "+collBoard.state)
        if(collBoard.get_oScreen().anyGreaterThan(1) == true){
            switch(key){
                case "a": this.left++; break;
                case "d": this.left--; break;
                case "s": this.top--; this.state = TetrisState.NewBlock; break;
                case "w":
                    this.idxBlockDegree = (this.idxBlockDegree!-1)%Tetris.nBlockDegrees;
                    this.currBlk = Tetris.setOfBlockObjects[this.idxBlockType!][this.idxBlockDegree];
                    break;
                case "_":
                    this.top--; this.state = TetrisState.NewBlock; console.log("new block needed \n\n"); break;
            }
            tempBlk = this.iScreen.clip(this.top,this.left,this.top+this.currBlk.get_dy(), this.left+this.currBlk.get_dx());
            tempBlk = tempBlk.add(this.currBlk);
            this.oScreen.paste(this.iScreen,0,0);
        }

        //

        if(this.state==TetrisState.Running){

            this.state = collBoard.accept(key)
            // super.accept(key)
            if(collBoard.get_oScreen().anyGreaterThan(1) != true ){
                // this.iScreen.paste(this.oScreen,0,0)
                var tmp:Matrix = this.iScreen.clip(collBoard.get_top(),collBoard.get_left(),collBoard.get_top()+collBoard.currBlk.get_dy(),collBoard.get_left()+collBoard.currBlk.get_dx())


                tmp = tmp.add(CTetris.setOfBlockObjects[collBoard.idxBlockType!][collBoard.idxBlockDegree!])

                this.oScreen.paste(this.iScreen,0,0)
                this.oScreen.paste(tmp, collBoard.get_top(),collBoard.get_left())

                // console.log("test")
                // this.drawMatrix(this.get_oScreen())
            }
        }
        else if(this.state==TetrisState.NewBlock){
                console.log("=============")

                // collBoard.accept(key)
                this.oScreen = this.deleteFullLines(this.oScreen, collBoard.currBlk, collBoard.top, this.iScreenDy,this.iScreenDx,Tetris.iScreenDw)
                // console.log("dfTest")
                // this.drawMatrix(this.oScreen)
                // this.oScreen = this.deleteFullLines(this.oScreen, collBoard.currBlk, collBoard.top, this.iScreenDy,this.iScreenDx,Tetris.iScreenDw)
                this.state= collBoard.accept(key);console.log("bBlock this.state:"+this.state)

                // if(collBoard.get_oScreen().anyGreaterThan(1) != true ){
                //         this.accept(key)
                // }
                // this.drawMatrix(this.get_oScreen());
                if(collBoard.get_oScreen().anyGreaterThan(1) != true ){

                        this.iScreen.paste(this.oScreen,0,0)
                        console.log("this.iScreen:")
                        this.iScreen.print()
                        if(key=="s"||key=="_"){
                            var tmp:Matrix = this.iScreen.clip(collBoard.get_top(),collBoard.get_left(),collBoard.get_top()+collBoard.currBlk.get_dy(),collBoard.get_left()+collBoard.currBlk.get_dx())
                            tmp = tmp.add(CTetris.setOfBlockObjects[collBoard.idxBlockType!][collBoard.idxBlockDegree!])

                            this.oScreen.paste(this.iScreen,0,0)
                            this.oScreen.paste(tmp, collBoard.get_top(),collBoard.get_left())
                            
                            

                        }



                        // console.log("test")
                        // this.drawMatrix(this.get_oScreen())
                }

        }


        // this.drawMatrix(this.oScreen)
        // this.oScreen.print()
        console.log("collBoard oScreen:")
        collBoard.oScreen.print()
        // console.log("here+++++++++++")
        // this.oScreen.print()
  

        return this.state;
    }
   

    drawMatrix(oScreen: Matrix): void {
        var dy:number = oScreen.get_dy();
        var dx:number = oScreen.get_dx();
        var array:any[][] = oScreen.get_array();

        for(var y=0; y<dy; y++){
            var oArray:string[]=[];
            for(var x=0; x<dx; x++){

                // if(array[y][x]==0) oArray.push("□ ")
                // else if (array[y][x]==1)  oArray.push("■ ")
                switch(array[y][x]){
                    case 0: oArray.push("□ "); break;
                    case 1: oArray.push("■ ");break;
                    case 10: oArray.push("♠ ");break;
                    case 20: oArray.push("★ ");break;
                    case 30: oArray.push("● ");break;
                    case 40: oArray.push("◆ ");break;
                    case 50: oArray.push("▲ ");break;
                    case 60: oArray.push("♣ ");break;
                    case 70: oArray.push("♥ ");break;
                    default: oArray.push("■ ");break;

                }
            }
            //console.log("\n");
            // console.log(array[y]);
            console.log(oArray);


            //230327


        }
    }


}//end of CTetris


export default CTetris;

File: src/main/kotlin/com/example/chat/model/CTetris.kt
package com.example.chat.model


class CTetris
    (cy:Int, cx:Int):Tetris(cy,cx){

    private var inboard = Tetris(15,10) // inboard에서 가져올 정보, top,left, currblk 모양
    private var back = Tetris(15,10)//배경정보(벽, 컬러블록 tmp제외)
    var valid:Boolean = true

    override fun deleteFullLines(screen: Matrix, blk: Matrix?, top: Int, dy: Int, dx: Int, dw: Int): Matrix {
        var line:Matrix; var zero:Matrix; var temp:Matrix?
        if(blk == null) return screen
        var cy:Int; var y:Int; var nDeleted:Int = 0; var nScanned:Int = blk.get_dy()
        if(top + blk.get_dy() - 1 >= dy)
            nScanned -= (top + blk.get_dy() - dy)
        zero = Matrix(1,dx)
        for(y in nScanned-1 downTo 0){
            cy = top + y + nDeleted
            line = screen.clip(cy,0,cy+1,screen.get_dx())
            if(line.int2bool().sum() == screen.get_dx()){
                temp = screen.clip(0,0,cy,screen.get_dx())
                screen.paste(temp,1,0)
                screen.paste(zero,0,dw)
                nDeleted++
            }
        }
        return screen
    }
    override fun printScreen(){
        var screen:Matrix = get_oScreen()
        var dy = screen.get_dy()
        println(dy)
        var dx = screen.get_dx()
        println(dx)
        var dw = get_iScreenDw()
        println(dw)
        var arr: Array<IntArray?>? = screen.get_array()
        var y:Int
        var x:Int
        for(y in 0 until dy-dw+1){
            for(x in dw-1 until dx-dw+1){
                if (arr!![y]!![x] == 0)
                    print("□  ")
                else if (arr[y]!![x] == 1)
                    print("■  ")
                else if (arr[y]!![x] == 10)
                    print("●  ")
                else if (arr[y]!![x] == 20)
                    print("♠  ")
                else if (arr[y]!![x] == 30)
                    print("▲  ")
                else if (arr[y]!![x] == 40)
                    print("▼  ")
                else if (arr[y]!![x] == 50)
                    print("♣  ")
                else if (arr[y]!![x] == 60)
                    print("◐  ")
                else if (arr[y]!![x] == 70)
                    print("♥  ")
            }
            println()
        }
    }

//    companion object{ //static과 유사한 역할
//        private var inboard = Tetris(15,10) // inboard에서 가져올 정보, top,left, currblk 모양
//        private var back = Tetris(15,10)//배경정보(벽, 컬러블록 tmp제외)
//    }
//<<<<<<< HEAD
//
//    private var inboard = Tetris(15,10) // inboard에서 가져올 정보, top,left, currblk 모양
//    private var back = Tetris(15,10)//배경정보(벽, 컬러블록 tmp제외)
//
//=======
//>>>>>>> 02980de91523e07d4bd571db2273721e414c51e0

    override fun accept(key: String): TetrisState {
        inboard.state = inboard.accept(key)
        var tmpblk = back.oScreen.clip(inboard.top, inboard.left,inboard.top+ inboard.currBlk!!.get_dy(), inboard.left+inboard.currBlk!!.get_dx())
        var type = inboard.idxBlockType+1
        var ncurrBlk = Matrix(inboard.currBlk!!)
        for(y in 0 until inboard.currBlk!!.get_dy()){
            for(x in 0 until inboard.currBlk!!.get_dx()){
                if (ncurrBlk.get_array()!![y]!![x] == 1)
                    ncurrBlk.get_array()!![y]!![x] = 10 * type
            }
        }
        oScreen = Matrix(back.oScreen) //배경화면에서 oscreen 따오기
        tmpblk = tmpblk.add(ncurrBlk) //tmpblk 생성
        oScreen.paste(tmpblk,inboard.top,inboard.left) // inboard연산 결과랑 동일한 위치에 tmpblk를 oscreen에 붙이기
        if(inboard.state == TetrisState.NewBlock){ //inboard state에서 새로운 블록이 필요하다면 oscreen에서도 deleteline 검사를 해줘야한다.
            oScreen = deleteFullLines(oScreen, ncurrBlk, inboard.top, inboard.iScreenDy,inboard.iScreenDx, inboard.get_iScreenDw())
            back.oScreen = Matrix(oScreen) //줄이 지워진다면 배경 정보 업데이트 해줘야함. 줄이 안지워져도 업데이트는 이루어져야한다.
        }
        state = inboard.state
        top = inboard.top
        left = inboard.left
        idxBlockType = inboard.idxBlockType
        idxBlockDegree = inboard.idxBlockDegree

        return state
    }
}